clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -151,7 +157,8 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
                             throw new KafkaException("Producer closed while allocating memory");
 
                         if (waitingTimeElapsed) {
-                            throw new TimeoutException("Failed to allocate memory within the configured max blocking time " + maxTimeToBlockMs + " ms.");
+                            this.metrics.sensor("buffer-exhausted-records").record();
+                            throw new BufferExhaustedException("Failed to allocate memory within the configured max blocking time " + maxTimeToBlockMs + " ms.");
```

Comment:
Since we mark the sensor here, there seems to be no need for the caller to know this is a BufferExhaustedException. Could we just throw TimeoutException and get rid of BufferExhaustedException?

Reply:
We could absolutely do that. 
I'd say that depends on whether it would be useful for code that uses the producer to distinguish between a timeout during metadata refresh and a timeout waiting for the buffer to free up enough space.
Personally I see no harm in allowing this distinction and since BufferExhaustedException extends TimeoutException no one is forced to do this, they can just catch both in one try block.

On the other hand I also see no huge benefit, as you would not treat these two any differently in code - your record wasn't sent, you either do something about it or ignore it. Unless you want to dynamically reconfigure your producer with a larger buffer when this occurred I don't see a real use-case. Maybe someone else can come up with one?

Reply:
Ok, we can keep this as its.